stetho/src/main/java/com/facebook/stetho/inspector/elements/android/ApplicationDescriptor.java

```diff
@@ -0,0 +1,155 @@
+
+  private ElementContext getContext(Application element) {
+    return mElementToContextMap.get(element);
+  }
+
+  @Override
+  protected void onHook(Application element) {
+    ElementContext context = new ElementContext();
+    context.hook(element);
+    mElementToContextMap.put(element, context);
+  }
+
+  @Override
+  protected void onUnhook(Application element) {
+    ElementContext context = mElementToContextMap.remove(element);
+    context.unhook();
+  }
+
+  @Override
+  protected NodeType onGetNodeType(Application element) {
+    return NodeType.DOCUMENT_NODE;
+  }
+
+  @Override
+  protected int onGetChildCount(Application element) {
+    ElementContext context = getContext(element);
+    return context.getActivitiesList().size();
+  }
+
+  @Override
+  protected Object onGetChildAt(Application element, int index) {
+    ElementContext context = getContext(element);
+    return context.getActivitiesList().get(index);
+  }
+
+  private final class ElementContext {
+    private Application mElement;
+    private Application.ActivityLifecycleCallbacks mCallbacks;
+    private List<Activity> mActivities;
+
+    public void hook(Application element) {
+      mElement = Util.throwIfNull(element);
+      mActivities = new ArrayList<>();
+
+      // TODO: tree diffing will remove the need to even worry about installing this callback,
+      //       which will then allow ~realtime updates for pre-ICS. for now, pre-ICS will just
+      //       have to close and re-open the inspector whenever the activity changes.
+      //       (this does assume that the hack below for getAllActivitiesHack() works on pre-ICS)
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.ICE_CREAM_SANDWICH) {
+        mCallbacks = new Application.ActivityLifecycleCallbacks() {
+          @Override
+          public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
+            mActivities.add(0, activity);
+            getListener().onChildInserted(mElement, null, activity);
+          }
+
+          @Override
+          public void onActivityStarted(Activity activity) {
+          }
+
+          @Override
+          public void onActivityResumed(Activity activity) {
+          }
+
+          @Override
+          public void onActivityPaused(Activity activity) {
+          }
+
+          @Override
+          public void onActivityStopped(Activity activity) {
+          }
+
+          @Override
+          public void onActivitySaveInstanceState(Activity activity, Bundle outState) {
+          }
+
+          @Override
+          public void onActivityDestroyed(Activity activity) {
+            mActivities.remove(activity);
+            getListener().onChildRemoved(mElement, activity);
+          }
+        };
+
+        mElement.registerActivityLifecycleCallbacks(mCallbacks);
+      }
+
+      try {
+        mActivities = getAllActivitiesHack();
+      } catch (Exception ex) {
```

Comment:
This should be encapsulated in the `...Hack` method.  Asking callers to deal with this is crazy.  For instance, what if there's an NPE in `getAllActivitiesHack`.  Do you really want the caller to have to figure that out?